### PR TITLE
Propagate protocol in block type

### PR DESCRIPTION
**Pull request by Jimbo4350**
This PR: 
- Propagates updated ShelleyBlock type (which is now also parameterized over the protocol)
- Re-enables the TestEnableDevelopmentHardForkEras flag so we can hardfork to Babbage without advertising we know about the Babbage era

### CODE REVIEW

**Review by MarcFontaine — APPROVED** [review]
LTGM

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 1]
```diff
@@ -193,13 +196,24 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
...
-
+        Praos.ProtocolParamsBabbage {
+          -- This is /not/ the Babbage protocol version. It is the protocol
+          -- version that this node will declare that it understands, when it
+          -- is in the Babbage era. Since Babbage is currently the last known
+          -- protocol version then this is also the Babbage protocol version.
+          Praos.babbageProtVer =
+            if npcTestEnableDevelopmentHardForkEras
+            then ProtVer 7 0  -- Advertise we can support Babbage
+            else ProtVer 6 0, -- Otherwise we only advertise we know about Mary,=
```
Comment (newhoggy) [inline cardano-node/src/Cardano/Node/Protocol/Cardano.hs:213]: Is the trailing `,=` intended?

**Review by newhoggy — APPROVED** [review]

**Comment by newhoggy** [comment]
This forks to Babbage:

```
$ CARDANO_NODE_SOCKET_PATH=example/node-bft1/node.sock cardano-cli query tip --testnet-magic 42
{
    "epoch": 0,
    "era": "Babbage",
    "syncProgress": "100.00"
}
```

**Comment by newhoggy** [comment]
But I noticed that `query tip` returns less data that Alonzo.  The following is what I get when only forking up to Alonzo:

```
$ CARDANO_NODE_SOCKET_PATH=example/node-bft1/node.sock cardano-cli query tip --testnet-magic 42
{
    "block": 2,
    "epoch": 0,
    "era": "Alonzo",
    "hash": "54787d3476ba5c9e9add8c1963c8a36f0f09b6271ef0b69fa775f6bdb3f003b4",
    "slot": 28,
    "syncProgress": "100.00"
}
```

File: cardano-api/src/Cardano/Api/InMode.hs [hunk 2]
```diff
@@ -185,25 +193,25 @@ toConsensusTxId (TxIdInMode txid ByronEraInCardanoMode) =
...
   txid' = Consensus.ShelleyTxId $ toShelleyTxId txid
 
 toConsensusTxId (TxIdInMode txid AlonzoEraInCardanoMode) =
   Consensus.HardForkGenTxId (Consensus.OneEraGenTxId (S (S (S (S (Z (Consensus.WrapGenTxId txid')))))))
  where
-  txid' :: Consensus.TxId (Consensus.GenTx (Consensus.ShelleyBlock Consensus.StandardAlonzo))
+  txid' :: Consensus.TxId (Consensus.GenTx Consensus.StandardAlonzoBlock)
   txid' = Consensus.ShelleyTxId $ toShelleyTxId txid
 
 toConsensusTxId (TxIdInMode _txid BabbageEraInCardanoMode) =
```
Comment (nc6) [inline cardano-api/src/Cardano/Api/InMode.hs:217]: Needs to be fixed at some point.

File: cardano-api/src/Cardano/Api/InMode.hs [hunk 3]
```diff
@@ -282,6 +290,10 @@ deriving instance Show (TxValidationErrorInMode mode)
 
 
 fromConsensusApplyTxErr :: ConsensusBlockForMode mode ~ block
+                        => Consensus.LedgerSupportsProtocol
+                             (Consensus.ShelleyBlock
+                             (TPraos.TPraos Consensus.StandardCrypto)
```
Comment (nc6) [inline cardano-api/src/Cardano/Api/InMode.hs:295]: Incorrectly fixed at `TPraos`

File: cardano-api/src/Cardano/Api/Modes.hs [hunk 4]
```diff
@@ -286,17 +290,27 @@ deriving instance Show (ConsensusModeParams mode)
...
-  ConsensusBlockForEra ShelleyEra = Consensus.ShelleyBlock StandardShelley
-  ConsensusBlockForEra AllegraEra = Consensus.ShelleyBlock StandardAllegra
-  ConsensusBlockForEra MaryEra    = Consensus.ShelleyBlock StandardMary
-  ConsensusBlockForEra AlonzoEra  = Consensus.ShelleyBlock StandardAlonzo
-
-
+  ConsensusBlockForEra ShelleyEra = Consensus.StandardShelleyBlock
+  ConsensusBlockForEra AllegraEra = Consensus.StandardAllegraBlock
+  ConsensusBlockForEra MaryEra    = Consensus.StandardMaryBlock
+  ConsensusBlockForEra AlonzoEra  = Consensus.StandardAlonzoBlock
```
Comment (nc6) [inline cardano-api/src/Cardano/Api/Modes.hs:301]: Missing a babbage case

File: cardano-node/src/Cardano/Node/Protocol/Cardano.hs [hunk 5]
```diff
@@ -218,6 +232,7 @@ mkSomeConsensusProtocolCardano NodeByronProtocolConfiguration {
                -- Version 3 is Allegra
                -- Version 4 is Mary
                -- Version 5 is Alonzo
+               -- TODO: I think Babbage is Version 7
```
Comment (nc6) [inline cardano-node/src/Cardano/Node/Protocol/Cardano.hs:235]: Yep

**Review by nc6 — APPROVED** [review]
I've left some comments on things that will need to be done. But in general, this looks good, and I'd be keen to merge this so we can begin testing a 1.35 version even without full Babbage support.

File: cardano-api/src/Cardano/Api/Block.hs [hunk 6]
```diff
@@ -238,7 +245,17 @@ fromConsensusBlock CardanoMode =
...
-toConsensusBlock :: ConsensusBlockForMode mode ~ block => BlockInMode mode -> block
+      Consensus.BlockBabbage b' ->
+        BlockInMode (ShelleyBlock ShelleyBasedEraBabbage b')
+                     BabbageEraInCardanoMode
+
+toConsensusBlock
+  :: ConsensusBlockForMode mode ~ block
+  => Consensus.LedgerSupportsProtocol
+       (Consensus.ShelleyBlock
+       (Consensus.TPraos Consensus.StandardCrypto)
```
[the remark this answers is not in the document]
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Block.hs:256]: Yep I need to fix this

File: cardano-api/src/Cardano/Api/InMode.hs (hunk 2 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/InMode.hs:217]: What's wrong with this?

File: cardano-api/src/Cardano/Api/InMode.hs (hunk 3 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/InMode.hs:295]: Yep need to fix

File: cardano-api/src/Cardano/Api/Block.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (nc6) [reply cardano-api/src/Cardano/Api/Block.hs:256]: Aha, sorry, I looked for `CardanoBlock` but missed the fixed era!

File: cardano-api/src/Cardano/Api/InMode.hs (hunk 3 above)
Comment (nc6) [reply cardano-api/src/Cardano/Api/InMode.hs:295]: If we're fixed to the Shelley era then that's fine. Sorry I didn't notice this!

**Review by deepfire — APPROVED** [review]

**Review by coot — APPROVED** [review]
LGTM

File: cabal.project [hunk 7]
```diff
@@ -191,12 +191,13 @@ source-repository-package
 source-repository-package
   type: git
   location: https://github.com/input-output-hk/cardano-ledger
-  tag: 1db68a3ec0a2dcb5751004beb22b906162474f23
-  --sha256: 03pv2jvskbi65dwaddp6a8bxbbcw674csjxhg8xbqd6q1kfpc41a
+  tag: 461506591729379168a607c3298746d0784621f1
```
Comment (kderme) [inline cabal.project:194]: Could we make sure this fix is included? https://github.com/input-output-hk/cardano-ledger/pull/2771

Comment (Jimbo4350) [reply cabal.project:194]: I'm also assuming we are waiting on the time bug fix as well?

Comment (kderme) [reply cabal.project:194]: Let's add the vrf check fix to the list.

**Comment by Jimbo4350** [comment]
bors merge

**Comment by disassembler** [comment]
bors merge

**Pull request merged**